design/fd-support-kcp.md

```diff
@@ -0,0 +1,111 @@
+# Failure Domain (FD) Support for Control-Plane Nodes
+
+## Summary
+
+CAPI supports failure domains (FDs) for control-plane nodes via the KCP
+controller. CAPI will look for the set of FailureDomains in the
+`ProviderCluster.Spec.FailureDomain`, if specified, the field will be copied
+to the `Cluster.Status.FailureDomains`. KCP will choose an FD from this set
+and place its value in `Machine.Spec.FailureDomain`. Currently, KCP tries
+to balance Machines evenly across all FDs.
+
+It is expected that providers will use this chosen FD (in the Machine.Spec)
+in deciding where to place the infrastructure machine. In the case of
+metal3, CAPM3 should associate the Metal3Machine with the corresponding BMH
+in the desired FD.
+
+In public clouds FailureDomains are well defined,. For example, in Azure, FDs
+map to availability zones within an azure region. For baremetal, there is no
+such mapping. However, an operator of a baremetal/metal3 cluster may still
+place their hosts across FDs in their data center or region.
+For example, two racks, connected to independent power grids, may
+constitue two FDs. Labels placed on the BMHs, by the operator, can provide a
+means to distinguish between hosts belonging to different FDs.
+
+The proposal is for CAPM3 to choose hosts based on the FD selected by KCP.
+In other words, based on the value in `Machine.Spec.FailureDomain` CAPM3 wil
+choose a BMH matching the label corresponding to that FD.
+
+## User Stories
+
+As an operator who has placed their baremetal infrastructure across different
+FDs, I would like CAPM3 to associate control-plane nodes with BMHs from the
+desired FD.
+
+## Goals
+
+* Support selection of BMHs based on the desired FD for the control plane node.
+
+## Non-Goals
+
+* Failure Domains support for worker nodes (i.e. MachineDeployments).
+
+## Proposal
+
+The operator labels the BMH resource with a standard FD label. For
+example, the following standard label could be used on the BMH:
+
+```diff
+  infrastructure.cluster.x-k8s.io/failure-domain=my-fd-1
```

Comment:
Is this a limitation of the CAPI Failure Domains model, that the failure domain is single-valued?

In the real world failure domains are hierarchical in nature - sometimes you want to spread hosts across multiple datacenters and sometimes just across adjacent racks. (See e.g. @jaypipes [here](http://lists.openstack.org/pipermail/openstack-dev/2014-March/031106.html) and "Partitioning and failure domains" on [this page](https://www.joinfu.com/2018/08/project-mulligans-architecture/).) To handle that in this design you'd need something like a separate controller (HWCC?) to check multiple labels and calculate the failure domain label. So it would be worth considering some kind of selector here, or allowing the user to choose (at the cluster level) the name of the label to match values for.

Reply:
Yes, at this point, CAPI only supports flat FDs. 

Kubernetes has the concept of zones and regions. Regions are composed of zone and thus hierarchical -- [K docs on this topic](https://kubernetes.io/docs/reference/labels-annotations-taints/#topologykubernetesioregion). A zone is essentially a failure domain within a region. While there is no strict definition of a region, major cloud providers consider regions to be geographically distributed. Additionally, creating a single K cluster that extends across regions is strongly discouraged. 

From a baremetal point of view, we can achieve hierarchy through the use of regions and zones -- for example, a region may represent a DC and a zone a rack (as in your example). In this scenario, a cluster most certainly can extend across "regions".  Unfortunately, this does not align with the cloud provider definition. 

In the near to mid-term, I think we'll only have zones to work with. Longer-term, there is a discussion to be had with the CAPI community in coming up with something that works well for both cloud and baremetal providers. 
By limiting ourselves to flat FDs, we can still mimic some hierarchy -- for example, as described in the [link that you gave](http://lists.openstack.org/pipermail/openstack-dev/2014-March/031106.html), you could make smaller FDs: rack1 -DC1, rack2-DC1, rack1-DC2,  rack2-DC2, etc… Not ideal but would be better than having no support. 

Also, please note that this proposal is only for control-plane nodes, so the initial scope is pretty limited.

Reply:
OK, if the limitation comes from CAPI then there is probably no current need in trying to design anything more complex. Thanks!